### PR TITLE
Pull: turn `.stream` and `.streamNoScope` into extension methods.

### PR DESCRIPTION
**Pull request by diesalbla**
The Pull methods `.scope` and `.streamNoScope` are currently implemented with the type-bounds evidence pattern and the
`asInstanceOf`. It may be better to define them as extension methods in an implicit class which, unlike the Pull class,
call be specialised to pulls where the result type is Unit.

This did not seem to require any special imports anywhere. It may be that, as implicits, they are attached to the `Pull` type everywhere.

### CODE REVIEW

File: core/shared/src/main/scala/fs2/Pull.scala [hunk 1]
```diff
@@ -147,6 +124,28 @@ sealed abstract class Pull[+F[_], +O, +R] {
 
 object Pull extends PullLowPriority {
 
+  implicit class StreamPullOps[F[_], O](val pull: Pull[F, O, Unit]) extends AnyVal {
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/Pull.scala:127]: Make `val` private or else this extension will allow syntax like `p.pull.pull.pull.pull.pull.pull`

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala:127]: `value class parameter must be a val and not be private[this]`

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala:127]: I can rename to `self`, if that helps...

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/Pull.scala]: `private val self: Pull[F, O, Unit]` should be fine... Note the other implicit classes we have with a public `self` val have the same issue -- `foo.self.self.self.self.self`

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala]: Ok... the problem is that this would mean allocating (you can no longer make them `AnyVal`).

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/Pull.scala]: That's not right...

```scala
io/src/main/scala-2.13/fs2/io/CollectionCompat.scala
27:  implicit class JIteratorOps[A](private val self: java.util.Iterator[A]) extends AnyVal {
30:  implicit class JIterableOps[A](private val self: java.lang.Iterable[A]) extends AnyVal {
33:  implicit class JSetOps[A](private val self: java.util.Set[A]) extends AnyVal {
36:  implicit class ListOps[A](private val self: List[A]) extends AnyVal {
39:  implicit class SetOps[A](private val self: Set[A]) extends AnyVal {
42:  implicit class EnumerationOps[A](private val self: java.util.Enumeration[A]) extends AnyVal {
```

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala]: 🤔

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala]: Ok, just pushed version with private and anyval. I got mistaken with the `private[this]` not allowed.

File: core/shared/src/main/scala/fs2/Pull.scala [hunk 2]
```diff
@@ -147,6 +124,28 @@ sealed abstract class Pull[+F[_], +O, +R] {
 
 object Pull extends PullLowPriority {
 
+  implicit class StreamPullOps[F[_], O](private val pull: Pull[F, O, Unit]) extends AnyVal {
```
Comment (nikiforo) [inline core/shared/src/main/scala/fs2/Pull.scala:127]: I looked on other random implicit classes in fs2, some of them have `final`, others don't.

Shouldn't final be added here?

Comment (diesalbla) [reply core/shared/src/main/scala/fs2/Pull.scala]: changed.

**Pull request merged**